src/config.cc

```diff
@@ -589,3 +590,39 @@ void Config::writeToFile(const boost::filesystem::path& filename) {
   writeOption(sink, import.tls_clientcert_path, "tls_clientcert_path");
   sink << "\n";
 }
+
+asn1::Serializer& operator<<(asn1::Serializer& ser, CryptoSource cs) {
+  ser << kAsn1Enum << static_cast<int32_t>(cs);
+
+  return ser;
+}
+
+asn1::Serializer& operator<<(asn1::Serializer& ser, const TlsConfig& tls_conf) {
+  ser << asn1::seq << kAsn1Utf8String << tls_conf.server << kAsn1Utf8String << tls_conf.server_url_path.string()
```

Comment:
@lbonn Yes, every "primitive" type (like string, integer, maybe bool in future) should be prepended with a type tag, the reason for that being that there are mutliple types in which a string/number can be serialized (like OCTET STRING, UTC TIME etc. for a string, INTEGER and ENUMERATED for an integer).

Reply:
Oh ok now I see. It looked different from the deserializer case because the operator is defined for `boost::filesysten::path` in that case.
Is there a reason for not defining `asn1::Serializer& operator<<(asn1::Serializer& ser, const boost::filesystem::path& path)`?

Do you think it would make sense to disallow and check for the chaining of two strings in row: `kAsn1Utf8String << str1 << str2`, to prevent errors?
Or removing the `std::string` instances and define a proxy type `Asn1Utf8String(std::string &s)` to disambiguate (and the stream operator would add the tag itself).

Reply:
@lbonn 
I like the idea of wrapper types, that will make the syntax more consistent. I'm not sure if exceptions on incorrect usage patters would be useful: we should probably see how often these errors will arise in practice.